### PR TITLE
Move preset logic from the controller to mutating webhook

### LINKED ISSUE #1278: Introduce a mutating webhook to set defaults and presets at creation/update time
**Is this a BUG REPORT or FEATURE REQUEST?**:

/kind enhancement

**What happened**:

Right now we have an extra controller which sets presets and defaults. This means that invalid specs can be persisted in the cluster because they are not necessarily complete on creation or update.

**What you expected to happen**:

Remove the preset controller and do presetting and defaulting on creation or update time.

**Anything else we need to know?**:

The MutatingAdmissionWebhook is beta since k8s 1,9, which means it is enabled by default. Save to adopt for us.

**Environment**:
- KubeVirt version (use `virtctl version`):
- Kubernetes version (use `kubectl version`):
- VM or VMI specifications:
- Cloud provider or hardware configuration:
- OS (e.g. from /etc/os-release):
- Kernel (e.g. `uname -a`):
- Install tools:
- Others:

### PR DESCRIPTION
**Pull request by cynepco3hahue**
**What this PR does / why we need it**:
We used the preset controller to apply presets and additional stuff(line namespace limit ranges, defaults...), but in this case, we did not have any validation after VMI creation.
Mutating webhook provide the more natural way to edit resource specification when a user creates or updates the resource.

- [x] validate specs on the update
- [x] set VMI defaults from mutating webhook
- [x] apply presets
- [x] apply cluster defaults
- [x] apply VMI foreground finalizer
- [x] add unittests
- [ ] adapt functional tests

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #1278

**Special notes for your reviewer**:
I think the PR close to the finish line, so I will appreciate any reviews :smiley: 

**Release note**:

```release-note
Move preset logic from the controller to mutating webhook
```

Signed-off-by: Artyom Lukianov <alukiano@redhat.com>

### CODE REVIEW

File: manifests/dev/rbac.authorization.k8s.yaml.in [hunk 1]
```diff
@@ -93,6 +93,7 @@ rules:
       - admissionregistration.k8s.io
     resources:
       - validatingwebhookconfigurations
+      - mutatingwebhookconfigurations
```
Comment (davidvossel) [inline manifests/dev/rbac.authorization.k8s.yaml.in]: release manifests need this as well.

File: pkg/virt-api/webhooks/mutating-webhook/mutating-webhook.go [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+	if err != nil {
+		return webhooks.ToAdmissionResponseError(err)
+	}
+
+	// Set VMI defaults
+	kubev1.SetObjectDefaults_VirtualMachineInstance(&vmi)
+
+	var patch []patchOperation
+	var value interface{}
+	value = vmi.Spec
```
Comment (davidvossel) [inline pkg/virt-api/webhooks/mutating-webhook/mutating-webhook.go]: Do we have any guarantee that the Mutation admission controller runs before the Validation one? If not, it's possible we should be running the resulting spec through the Validation logic at the end of this.

File: manifests/dev/rbac.authorization.k8s.yaml.in (hunk 1 above)
Comment (cynepco3hahue) [reply manifests/dev/rbac.authorization.k8s.yaml.in]: @davidvossel I wanted to ask you about it because I did not find any `validatingwebhookconfigurations` under release manifests, does its part of some general permissions? Also, I can see that release functional tests passed the stage of deployment.

File: pkg/virt-api/webhooks/mutating-webhook/mutating-webhook.go (hunk 2 above)
Comment (rmohr) [reply pkg/virt-api/webhooks/mutating-webhook/mutating-webhook.go]: Yes. It's guaranteed that the mutations run before the validation.

Comment (cynepco3hahue) [reply pkg/virt-api/webhooks/mutating-webhook/mutating-webhook.go]: @davidvossel Yes, we have guarantees(good for us :smile: )
https://docs.okd.io/latest/architecture/additional_concepts/dynamic_admission_controllers.html

**Comment by cynepco3hahue** [comment]
1. I like it, but we still will have the same problem when a user will update a VM object. @davidvossel Do you know if we can configure k8s to apply webhooks also on an object update?
Currently, I can see that if I will update a VM template with some illegal value and will start the VM, I will see the validation error only under `virt-controller` log(VMI one).
2. I think it absolutely valid operation from a user perspective to update a VM preset(or template) after creation, so I do not think that we need to restrict it, but I do not sure how we can validate it(maybe we can forward admissionResponseError from VMI to VM).

**Comment by rmohr** [comment]
> @cynepco3hahue hey, not sure what is our approach here, but we don't need to implement the whole admission server, we can use a framework hosted under openshift repo https://github.com/openshift/generic-admission-server/. I used it for my mutating webhook and it works well.

Can't say if I am for or against it. But let's keep in mind that they also vendor k8s dependencies. It may get hard for us to synchronize k8s dependencies in the vendor folder.

**Comment by fabiand** [comment]
Please make the title more descriptive.

"Add mutatoing webhook in order to …"

**Comment by fabiand** [comment]
Please also make the release note more descriptive.

File: manifests/dev/rbac.authorization.k8s.yaml.in (hunk 1 above)
Comment (davidvossel) [reply manifests/dev/rbac.authorization.k8s.yaml.in]: umm, that's not good. I'll take a look

**Comment by cynepco3hahue** [comment]
@fabiand Will update the title, description and release note when we will decide what future course of this change regarding presets.

**Comment by davidvossel** [comment]
> hey, not sure what is our approach here, but we don't need to implement the whole admission server, we can use a framework hosted under openshift repo https://github.com/openshift/generic-admission-server/. I used it for my mutating webhook and it works well.

I'm hesitant to pull in this dependency when there's not much code for us to leverage there. This might not justify the maintenance required to keep the dependency in sync even.

**Comment by davidvossel** [comment]
>  k8s to apply webhooks also on an object update

@cynepco3hahue good point. Validation webhooks can occur on update. However, I just checked and we're only registering validation webhooks for the create operation. We need to add update and patch to the operation list there.

**Comment by rmohr** [comment]
> need to add update and patch to the operation list there.

And update and patch should basically not allow any changes to the spec section since right now the vmis are immutable.

**Comment by cynepco3hahue** [comment]
> And update and patch should basically not allow any changes to the spec section since right now the vmis are immutable.

Only in VMI case, we still have VM and VMIRS.

File: manifests/dev/rbac.authorization.k8s.yaml.in (hunk 1 above)
Comment (davidvossel) [reply manifests/dev/rbac.authorization.k8s.yaml.in]: @cynepco3hahue  i figured out why. We have issues with our RBAC rules. #1415 will fix it. We'll need the mutation webhook privilege for virt-api in both once my PR gets merged

**Comment by cynepco3hahue** [comment]
Looks like that I will need to create a separate endpoint for the VMI update, because in both cases(create or update), request use POST method, so I can not differentiate them.

**Comment by cynepco3hahue** [comment]
@davidvossel I think instead of introducing presets for VM's and VMIRS's, I will improve validation.
Because if we will apply preset on VM template spec, it already will be part of the VM object and not of a preset object that can be a problem if a user will want to replace presets.
So my plan is, to apply presets(if exists) to a template spec, create dummy VMI object from this spec and send it to `ValidateVirtualMachineInstanceSpec`, I think it will get as the desired behavior.

**Comment by davidvossel** [comment]
> I think instead of introducing presets for VM's and VMIRS's, I will improve validation.
Because if we will apply preset on VM template spec, it already will be part of the VM object and not of a preset object that can be a problem if a user will want to replace presets.
So my plan is, to apply presets(if exists) to a template spec, create dummy VMI object from this spec and send it to ValidateVirtualMachineInstanceSpec, I think it will get as the desired behavior.

This still won't catch the scenario where a preset value changes between the time a VM is posted and the VMI is started. 

What I'm hung up on with all of this preset logic is that it's possible for a VM to start VMIs with different specs between runs if a VMIPreset changes.

Coming from a k8s background, that concept isn't strange to me. For example, if my Deployment is spinning up Pods and I change a PodPreset, then I'd expect new pods being created in my Deployment to pick up the preset change if it matches.

However, coming from a VM background this gets weird. If I stop a VM then start it again, I'd expect the exact same behavior... but with VMIPresets we're saying that between starting and stopping a VM, something can change. I've never seen this kind of behavior before in the VM world and it would surprise me as a user.

@rmohr @fabiand do either of you have thoughts on this?

**Comment by cynepco3hahue** [comment]
@davidvossel hm I see, currently, we do not spawn new VMI from the VM on the VM template spec change, it will need some additional logic under VM controller and we need to consider if we want it at all, so let's put it aside.
About validation, on the preset update, I can check all VM's and VMIRS's that uses this preset(do not sure how resource consumable will it be on the large scale, also with cache usage), under the validation I will apply new preset on the VM template spec and validate it, in case of error I will deny updating the preset.

**Comment by fabiand** [comment]
First: Presets should only be applied to VMIs, not VM's VMI templates.

> What I'm hung up on with all of this preset logic is that it's possible for a VM to start VMIs with different specs between runs if a VMIPreset changes.

That at least makes a validation during VM creation moot, but in general this behavior is as intended, as in that a VMI will "receive new updates done to it's corresponding preset".

IIUIC the goal is to validate the VM during creation, including the VMI template, including tany potential preset.
First: I'd check what Kubernetes does with PodPresets, if they get validated during i.e. Deployment creation. I'd doubt that they do this.

However, what again becomes interesting in this context is a dry-run for VMIs.
This would allow a user or (UI) system to create the VM, then use the dry-run in order to understand the final VMI - including it's validation and getting error feedback.

> Coming from a k8s background, that concept isn't strange to me. For example, if my Deployment is spinning up Pods and I change a PodPreset, then I'd expect new pods being created in my Deployment to pick up the preset change if it matches.

It's worth a discussion to see if this should apply to VMs, or rather to VMIRS like controllers.

> I've never seen this kind of behavior before in the VM world and it would surprise me as a user.

I'd also not expect the "principal" presets to be volatile or changing often.
I.e. modifying a preset could be comparable to modify an instance type in OS.

**Comment by rmohr** [comment]
> That at least makes a validation during VM creation moot, but in general this behavior is as intended, as in that a VMI will "receive new updates done to it's corresponding preset".

We should limit presets to cases where the VMI template standalone is already valid and and the preset itself is also a valid subset, based on the extension points which we allow the presets to expand.

Considering how k8s and kubevirt works, templates or other solutions at the same level (before creating a VM or VMI) are a better fit for representing nont-necessarily already valid VMs and VMIs.

> 
> IIUIC the goal is to validate the VM during creation, including the VMI template, including tany potential preset.
> First: I'd check what Kubernetes does with PodPresets, if they get validated during i.e. Deployment creation. I'd doubt that they do this.

Only validated on creaton. Presets can be changed anytime, so even if the validation would fail right now, it might work later when you actually create the Pod.

> 
> However, what again becomes interesting in this context is a dry-run for VMIs.
> This would allow a user or (UI) system to create the VM, then use the dry-run in order to understand the final VMI - including it's validation and getting error feedback.

Having a dry-run query parameter is also common in k8s.

**Comment by rmohr** [comment]
> This still won't catch the scenario where a preset value changes between the time a VM is posted and the VMI is started.
> 
> What I'm hung up on with all of this preset logic is that it's possible for a VM to start VMIs with different specs between runs if a VMIPreset changes.
> 
> Coming from a k8s background, that concept isn't strange to me. For example, if my Deployment is spinning up Pods and I change a PodPreset, then I'd expect new pods being created in my Deployment to pick up the preset change if it matches.
> 
> However, coming from a VM background this gets weird. If I stop a VM then start it again, I'd expect the exact same behavior (same spec used)... but with VMIPresets we're saying that between starting and stopping a VM, something can change. I've never seen this kind of behavior before in the VM world and it would surprise me as a user.

In general I share your concerns. I would still want to keep presets where they are and limit their capabilities to preset fields to not-strictly required fields (e.g. not memory). Mostly the features section, disks and networks are good fits.

I see presets as a way to configure common parts in a way so that they are explicitly picked up if they are changed. For the other cases I think templates or other mechanism on the same level where templates act  are a better fit.

**Comment by cynepco3hahue** [comment]
@davidvossel @rmohr @fabiand I do not sure that I want to make this PR bigger than it now, so I prefer to be sure that all preset logic continue work as with controller and additional features leave to other PR's.
Things that we need to consider

**Comment by fabiand** [comment]
I think that's always good. One PR for one thing :)

**Comment by davidvossel** [comment]
> @davidvossel @rmohr @fabiand I do not sure that I want to make this PR bigger than it now, so I prefer to be sure that all preset logic continue work as with controller and additional features leave to other PR's.

yep, sorry for getting us side tracked. There's nothing blocking us from moving the existing preset logic to mutation webhooks. 

@cynepco3hahue for the scope of this PR, your approach looks good.  I'll give it another detailed review today and we'll move this forward as is.

File: tests/vmi_lifecycle_test.go [hunk 3]
```diff
@@ -129,21 +129,6 @@ var _ = Describe("VMIlifecycle", func() {
 				500*time.Millisecond).
 				Should(ContainSubstring("info : hostname: " + dns.SanitizeHostname(vmi)))
 		})
-
-		It("should reject POST if schema is invalid", func() {
```
Comment (davidvossel) [inline tests/vmi_lifecycle_test.go]: why was this removed?

File: pkg/api/v1/types.go [hunk 4]
```diff
@@ -646,14 +646,14 @@ type VirtualMachineInstancePresetSpec struct {
 	// Required.
 	Selector metav1.LabelSelector `json:"selector"`
 	// Domain is the same object type as contained in VirtualMachineInstanceSpec
-	Domain *DomainPresetSpec `json:"domain,omitempty"`
+	Domain *DomainSpec `json:"domain,omitempty"`
```
Comment (davidvossel) [inline pkg/api/v1/types.go]: it's weird that the preset spec didn't originally use the DomainSpec directly, especially since they contain all the same values.

@stu-gott Was there a reason for the separate DomainPresetSpec?

File: manifests/dev/rbac.authorization.k8s.yaml.in [hunk 5]
```diff
@@ -125,6 +126,21 @@ rules:
     verbs:
       - get
       - list
+      - watch
+  - apiGroups:
+      - ''
+    resources:
+      - limitranges
```
Comment (davidvossel) [inline manifests/dev/rbac.authorization.k8s.yaml.in]: does virt-controller no longer need access to limitranges now? If so it seems like we could remove that permission from the virt-controller ClusterRole

Comment (cynepco3hahue) [reply manifests/dev/rbac.authorization.k8s.yaml.in]: done

File: tests/vmi_lifecycle_test.go (hunk 3 above)
Comment (cynepco3hahue) [reply tests/vmi_lifecycle_test.go]: this one interesting, so the scenario is:
- defines new VMI
- changes `domain` under spec to `not-a-domain`
- creates VMI, but first, it sends VMI to mutating webhook that set defaults, so it will add `domain` section to VMI, and now under VMI spec we have two sections. `domain` and `not-domain`, so API server the is valid(it just ignore `not-domain` section)
I tried to figure out how can save this test, but I lack of ideas.

Comment (davidvossel) [reply tests/vmi_lifecycle_test.go]: Ha, that's funny. I'm fine with just removing the test then.

File: tests/vmidefaults_test.go [hunk 6]
```diff
@@ -67,10 +69,11 @@ var _ = Describe("VMIDefaults", func() {
 
 		It("Should be applied to VMIs", func() {
 			// create the VMI first
-			err = virtClient.RestClient().Post().Resource("virtualmachineinstances").Namespace(tests.NamespaceTestDefault).Body(vmi).Do().Error()
+			_, err = virtClient.VirtualMachineInstance(tests.NamespaceTestDefault).Create(vmi)
```
Comment (davidvossel) [inline tests/vmidefaults_test.go]: you should actually be able to see the changes right here in the returned VMI.

```newVmi, err := virtClient.VirtualMachineInstance(tests.NamespaceTestDefault).Create(vmi)```

that's all that is needed now. You can get rid of the GET below.

File: tests/vmipreset_test.go [hunk 7]
```diff
@@ -170,10 +170,12 @@ var _ = Describe("VMIPreset", func() {
 
 			newPreset := waitForPreset(virtClient, memoryPrefix)
 
-			err = virtClient.RestClient().Post().Resource("virtualmachineinstances").Namespace(tests.NamespaceTestDefault).Body(vmi).Do().Error()
+			_, err = virtClient.VirtualMachineInstance(tests.NamespaceTestDefault).Create(vmi)
 			Expect(err).ToNot(HaveOccurred())
+			tests.WaitForSuccessfulVMIStart(vmi)
 
-			newVMI := waitForVirtualMachine(virtClient)
+			newVMI, err := virtClient.VirtualMachineInstance(tests.NamespaceTestDefault).Get(vmi.Name, &k8smetav1.GetOptions{})
```
Comment (davidvossel) [inline tests/vmipreset_test.go]: like the other test I pointed out, you shouldn't need a GET here. The newVMI can be returned in the Create with all the changes... or at least that's supposed to be the way it works :D

File: tests/vmipreset_test.go [hunk 8]
```diff
@@ -190,13 +192,15 @@ var _ = Describe("VMIPreset", func() {
...
+			vmi = tests.NewRandomVMIWithEphemeralDisk(tests.RegistryDiskFor(tests.RegistryDiskAlpine))
 			vmi.Labels = map[string]string{flavorKey: cpuFlavor}
 
-			err = virtClient.RestClient().Post().Resource("virtualmachineinstances").Namespace(tests.NamespaceTestDefault).Body(vmi).Do().Error()
+			_, err = virtClient.VirtualMachineInstance(tests.NamespaceTestDefault).Create(vmi)
 			Expect(err).ToNot(HaveOccurred())
+			tests.WaitForSuccessfulVMIStart(vmi)
 
-			newVMI := waitForVirtualMachine(virtClient)
+			newVMI, err := virtClient.VirtualMachineInstance(tests.NamespaceTestDefault).Get(vmi.Name, &k8smetav1.GetOptions{})
```
Comment (davidvossel) [inline tests/vmipreset_test.go]: shouldn't need GET

File: tests/vmipreset_test.go [hunk 9]
```diff
@@ -216,11 +220,13 @@ var _ = Describe("VMIPreset", func() {
...
 			// reset the label so it will not match
-			vmi = tests.NewRandomVMIWithEphemeralDisk("kubevirt/alpine-registry-disk-demo:devel")
-			err = virtClient.RestClient().Post().Resource("virtualmachineinstances").Namespace(tests.NamespaceTestDefault).Body(vmi).Do().Error()
+			vmi = tests.NewRandomVMIWithEphemeralDisk(tests.RegistryDiskFor(tests.RegistryDiskAlpine))
+			_, err = virtClient.VirtualMachineInstance(tests.NamespaceTestDefault).Create(vmi)
 			Expect(err).ToNot(HaveOccurred())
+			tests.WaitForSuccessfulVMIStart(vmi)
 
-			newVMI := waitForVirtualMachine(virtClient)
+			newVMI, err := virtClient.VirtualMachineInstance(tests.NamespaceTestDefault).Get(vmi.Name, &k8smetav1.GetOptions{})
```
Comment (davidvossel) [inline tests/vmipreset_test.go]: shouldn't need GET

File: tests/vmipreset_test.go [hunk 10]
```diff
@@ -232,12 +238,14 @@ var _ = Describe("VMIPreset", func() {
 
 		It("Should not be applied to existing VMIs", func() {
 			// create the VirtualMachineInstance first
-			err = virtClient.RestClient().Post().Resource("virtualmachineinstances").Namespace(tests.NamespaceTestDefault).Body(vmi).Do().Error()
+			_, err = virtClient.VirtualMachineInstance(tests.NamespaceTestDefault).Create(vmi)
 			Expect(err).ToNot(HaveOccurred())
+			tests.WaitForSuccessfulVMIStart(vmi)
 
-			newVMI := waitForVirtualMachine(virtClient)
+			newVMI, err := virtClient.VirtualMachineInstance(tests.NamespaceTestDefault).Get(vmi.Name, &k8smetav1.GetOptions{})
```
Comment (davidvossel) [inline tests/vmipreset_test.go]: shouldn't need GET

File: tests/vmipreset_test.go [hunk 11]
```diff
@@ -257,15 +265,17 @@ var _ = Describe("VMIPreset", func() {
...
 			exclusionMarking := "virtualmachineinstancepresets.admission.kubevirt.io/exclude"
 			vmi.Annotations = map[string]string{exclusionMarking: "true"}
 
-			err = virtClient.RestClient().Post().Resource("virtualmachineinstances").Namespace(tests.NamespaceTestDefault).Body(vmi).Do().Error()
+			_, err = virtClient.VirtualMachineInstance(tests.NamespaceTestDefault).Create(vmi)
 			Expect(err).ToNot(HaveOccurred())
+			tests.WaitForSuccessfulVMIStart(vmi)
 
-			newVMI := waitForVirtualMachine(virtClient)
+			newVMI, err := virtClient.VirtualMachineInstance(tests.NamespaceTestDefault).Get(vmi.Name, &k8smetav1.GetOptions{})
```
Comment (davidvossel) [inline tests/vmipreset_test.go]: shouldn't need GET

**Comment by cynepco3hahue** [comment]
ci test please

**Comment by cynepco3hahue** [comment]
We have a lot of failures on fedora repository
```
Step 3/11 : RUN dnf -y install   socat   genisoimage   util-linux   libcgroup-tools   ethtool   net-tools   sudo && dnf -y clean all &&   test $(id -u qemu) = 107 # make sure that the qemu user really is 107
 ---> Running in a7c983ae1435

[91m[0mFedora 28 - x86_64 - Updates                    2.1 MB/s |  24 MB     00:11    
Virtualization packages from Rawhide built for  110 kB/s |  57 kB     00:00    
[91mError: Failed to synchronize cache for repo 'fedora'
```

**Comment by cynepco3hahue** [comment]
Tests look much better, so I will merge this one.
